### DISCUSSION TITLE
Native Mobile App

### DISCUSSION THREAD

**Discussion by hrueger** [josephdadams/TallyArbiter#451] [Ideas]
As @josephdadams said in #450, it would be cool to have a native app for mobile devices.
I thought I'd share my experiences and thoughts about the technologies I've tried so far:

**Apache Cordova / Phonegap**
This is basically just a wrapper for a web view displaying some html. It offers some JS APIs to access native things like the camera or the acceleration sensor data. However, as it is just a webview, it does not feel very native in my opinion. Also, most of the features can now be used in a normal web browser, so a PWA is just the same.

**NativeScript**
This is a framework for hybrid apps allowing us to write our UI in some form of XML which is then rendered as native elements. The look (which is already great since the UI elements are native elements) can be customized using (S)CSS and the logic can be in TypeScript or JavaScript. It supports VanillaJS as well as other frameworks like Vue or Angular.
In theory it also supports code sharing (same logic for the web and nativescript app and you only need to write the UI code twice) but I've had a lot of trouble in the passt getting this to work and even more trouble to make it still work after platform updates.

**Native**
Here I can only talk about Android App development as I don't own an iPhone. 
Writing an Android App is not that hard, especially now with Kotlin. If we strive for the best performance and best look and feel, I think there's no way around native apps.
However, the drawback is that we'd have to build and maintain two codebases, one for iOS and one for Android.

**Flutter**
In my opinion this would be the best way. As far as I know, the dart language is compiled and not interpreted like with the JS in NativeScript. This should give us better performance.
I really like the way layouts are done in Flutter and as it compiles to Android or iOS Apps, we'd only have a single codebase for the mobile app.
I do have to confess though, that I've not built more than a simple hello world app with it yet.

What do you think? Do you have any experiences with app development?

**Comment by josephdadams** [comment]
Unless we had a developer jump in who wanted to maintain a native codebase for iOS, I think Flutter would be the best option. I think creating the simple web tally functionality would be a great place to start and would hopefully be straightforward to implement.

I don't know much about Android apps since I live in iOS world but I do know that with iOS you have to pay a yearly fee to have your apps signed and listed in the store.

**Reply by commputethis** [reply]
Android is just a flat one time fee.  I haven't looked for years, but when I got my developer account, it was $25.  I just created a basic calculator app for nutrition conversion for my wife and I to use for calculating carbs for my daughter who has Type 1 Diabetes.  I wanted to put it in the store so it was a "trusted source."  I guess I say that to say that I am not real good with creating apps, but I could potentially help manage it.  I mean, I love the Tally Arbiter software and would like to give back how I can.  I am just not confident enough to create something.

**Reply by commputethis** [reply]
I just went and checked.  Android Developer account is still just a one time $25 fee.  Apple Developer account is $99/year.

**Reply by josephdadams** [reply]
@commputethis I welcome anything you want to contribute :)

**Reply by commputethis** [reply]
I would be willing to cover the Android Developer fee and $25 towards Apple Developer if others would pitch in to do the rest.

I did install Flutter last night and working my way through their tutorial, but, again, I'm not real confident how much I can contribute to designing an app.  @hrueger, what is it we want the app to do?

**Reply by josephdadams** [reply]
Ideally it would include the features of the "web tally" listener client - and keep the phone screen alive while running, support vibration, chat, flashing, etc. All the features of the web tally but in a better experience for the user. MDNS detection of TA servers could be nice, built-in QR code scanning to hop to a specific server, etc.

**Comment by logandavid1** [comment]
Flutter is the best app development solution as you do not need to write in a different codebase for each platform. Above all, it has rich libraries and fast testing support.
